packages/core/src/codewhisperer/service/serviceContainer.ts

```diff
@@ -0,0 +1,33 @@
+/*!
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+import { AuthUtil } from '../util/authUtil'
+import { ActiveStateController } from '../views/activeStateController'
+import { LineTracker } from '../tracker/lineTracker'
+
+export class Container {
+    static #instance: Container | undefined
+
+    static create(authProvider: AuthUtil): Container {
+        if (Container.#instance) {
+            throw new Error('Container already exists')
+        }
+
+        Container.#instance = new Container(authProvider)
+        return Container.#instance
+    }
+
+    static get instance(): Container {
+        return Container.#instance ?? Container.create(AuthUtil.instance)
+    }
+
+    readonly lineTracker: LineTracker
+    readonly activeStateController: ActiveStateController
```

Comment:
yea, agree, that's the reason why I add this container class, afterward, I want to bring in other components into this container so that they can have reference to each other easily and don't need to make them singleton

Reply:
so by passing the container into its children components' constructors, we can just `this.container.otherCompoenent` instead of `otherCompoenent.getInstance`